soc: kinetis: full pinctrl enablement for Kinetis boards

danieldegrasse opened the pull request:
This PR enables pinctrl for all drivers used by NXP Kinetis SOCs, and disables pinmux drivers for in tree Kinetis boards. The in tree pinmux driver is still present, but is no longer enabled at the SOC level. All in tree Kinetis boards should now be converted for pinctrl support.

Dependent on #42238 for pinctrl support in the MCUX LPUART driver.

danieldegrasse committed:
drivers: pwm: Enable pinctrl for kinetis ftm PWM driver

Enable pinctrl for kinetis ftm PWM driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove ftm pinmux configuration

Remove ftm pinmux configuration from kinetis boards, as ftm driver
supports pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: i2c: enable pinctrl for kinetis i2c

Enable pinctrl for kinetis i2c mcux driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove i2c pinmux for kinetis boards

Remove i2c pinmux for kinetis boards, as i2c driver supports pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: spi: enable pinctrl for kinetis dspi driver

Enable pinctrl unconditionally for kinetis dspi driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove spi pinmux setting for kinetis boards

Remove spi pinmux setting for kinetis boards with dspi, as dspi driver
supports pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: enet: Enable pinctrl for eth_mcux driver

Enable pinctrl for ethernet mcux driver, and update kinetis DTS node to
include labelling for PTP node, to enable driver to access pinctrl
properties.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove enet pinmux for kinetis boards

Remove unnecessary usage of pinmux driver for kinetis boards with
ethernet

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: can: Enable pinctrl for mcux_flexcan

Enable pinctrl support for mcux flexcan driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove flexcan pinmux configuration for kinetis boards

Remove flexcan pinmux configuration for kinetis boards, as flexcan
driver uses pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: adc: enable pinctrl for kinetis adc_16 mcux driver

Enable pinctrl for kinetis adc16 mcux driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove unused adc pinmux from kinetis boards

adc16 driver supports pinctrl, so remove unused pinmux from kinetis
boards

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: set CONFIG_PINCTRL=y for all kinetis boards

Set CONFIG_PINCTRL=y for all kinetis boards, so they can use the
relevant drivers that enable pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove lpuart pinmux setting for kinetis boards

Remove lpuart pinmux setting for all kinetis boards, as lpuart driver
supports pinctrl.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: i2c: enable pinctrl for mcux lpi2c driver

Enable pinctrl for mcux lpi2c driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove lpi2c pinmux usage in kinetis boards

remove pinmux for lpi2c devices on kinetis boards, as these boards
support pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: spi: enable pinctrl for spi_mcux_lpspi

Enable pinctrl for lpspi driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: twr_ke18f: Move lpspi to pinctrl

LPSPI peripheral driver supports pinctrl. Move twr_ke18f pinmux to use
pinmux for lpspi, and apply dynamic pinctrl states in order to select
correct chip select pin.

Also add pinmux settings to LPSPI for RT1060, so that LPSPI peripheral
driver will continue to work.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pwm: add pinctrl support for mcux PWT driver

Enable pinctrl support for mcux pwm capture pwt driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove pinmux configuration for kinetis boards using pwt

Remove pinmux configuration for kinetis boards using pwt, as this driver
supports pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: adc: adc_mcux_adc12: enable pinctrl for adc12 driver

enable pinctrl for adc12 driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove pinmux configuration for kinetis boards using adc12

Remove pinmux configuration for kinetis boards using adc12, as pinctrl
is supported by driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: sensor: Add pinctrl support for mcux acmp sensor

Add pinctrl support for mcux acmp sensor driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove pinmux usage for acmp peripheral

Remove pinmux usage for acmp peripheral on kinetis boards, as mcux_acmp
driver supports pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: dac: add pinctrl support to dac_mcux_dac32 driver

add pinctrl support to dac_mcux_dac32 driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove pinmux usage for dac32 driver

dac32 driver supports pinctrl. Remove pinmux usage on kinetis boards
with this IP.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pwm: add pinctrl support to kinetis tpm driver

add pinctrl support for kinetis tpm driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove tpm pinmux usage for kinetis boards

tpm driver supports pinctrl. Remove pinmux usage for kinetis boards.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
boards: remove pinmux usage for all kinetis boards

Remove pinmux file where possible, and remove all pinmux driver usage
where file cannot be removed. All kinetis boards no longer use pinmux
driver.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: disable pinmux driver for nxp kinetis SOCs

disable pinmux driver for nxp kinetis SOCs, since all boards use pinctrl

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: kinetis: do not redefine PORT_PCR_ODE for KW24/22 series

KW24 and KW22 series support open drain pins, while rest of kinetis KW
SOCs do not. Don't redefine PORT_PCR_ODE for these SOCs

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
soc: kw2x: Fix pinctrl definition for SPI modem

Fix pinctrl definition for SPI modem for kw2x SOCs

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@hakehuang Could you run CI on this PR, specifically targeting Kinetis boards? I want to make sure all the pinmux settings are correct.

hakehuang commented:
@danieldegrasse  CI is running will feedback tomorrow.

henrikbrixandersen commented on drivers/can/can_mcux_flexcan.c:
```diff
@@ -702,6 +709,13 @@ static int mcux_flexcan_init(const struct device *dev)
 	}
 
+#ifdef CONFIG_PINCTRL
+	err = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
```
I think `pinctrl_apply_state()` should go before `mcux_flexcan_set_mode()` to ensure the pins are routed to the CAN transceiver before trying to initialize the CAN controller?

danieldegrasse replied:
thanks for pointing that out, I've moved the `pinctrl_apply_state` to before `mcux_flexcan_set_mode`

henrikbrixandersen reviewed (changes requested):
Looks good, thank you. I know it's still in draft, but I took the liberty to review it anyway :)

I noticed that in some drivers you conditionally use pinctrl, others you always use it regardless if it is enabled or not. I take it the conditional ones are due to the same drivers being in use in other SoC families (e.g. i.MX RT)?

Anyways, for drivers that unconditionally use the pinctrl API, I think the Kconfig symbols needs to depend on `CONFIG_PINCTRL`?

hakehuang commented:
@danieldegrasse on frdm_k64f I meet two driver test error. 

1. tests/driver/adc. failure log is 
[adc_failure_log.txt](https://github.com/zephyrproject-rtos/zephyr/files/8223595/adc_failure_log.txt)
2. tests/driver/spi/spi_loopback. failure log is
[spi_failure_log.txt](https://github.com/zephyrproject-rtos/zephyr/files/8223588/spi_failure_log.txt)

for spi case, you need short the sin and sout which are: ptd2 - ptd3

the two above cases are OK in mainline weekly testing which is zephyr-v3.0.0-838-gc050d1f2ec.

danieldegrasse commented:
@henrikbrixandersen Thanks for the feedback. That is why some drivers have conditional pinctrl enablement- once full iMX.RT support is merged, that should be dropped.

As far as the Kconfig symbols, I think that rather than depending on `CONFIG_PINCTRL` the drivers should select `CONFIG_PINCTRL`. This way, a new user to the project doesn't need to debug Kconfig symbols to understand why a driver isn't being included. With this, I will update the corresponding dts binding so that `pinctrl-0` nodes are required. This way a user should get reasonably useful build errors if their board does not support pinctrl

danieldegrasse committed:
drivers: pinctrl: Fix kinetis pinctrl slew rate

kinetis pinctrl driver had swapped values for slew rate
(fast slew rate should set bit to zero). Fix slew rate values.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

hakehuang commented:
> @hakehuang I reproduced and fixed the ADC test failure, but I cannot reproduce the SPI failure. I tried testing with and without the `mcux-dspi-dma.conf` overlay applied, and in both cases the test passed when ptd2 and ptd3 were shorted. However, the pinctrl driver did have an error regarding the setting of the slew rate (`SRE` field), where it incorrectly set `SRE=1` when a fast slew rate was selected, so I've fixed that.

I will re-run CI based your updates, and feedback

hakehuang commented:
> I will re-run CI based your updates, and feedback

ALl CI PASS.

danieldegrasse committed:
drivers: serial: Enable pinctrl for kinetis lpsci driver

Enable pinctrl for kinetis lpsci uart driver

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
@hakehuang I have updated the LPSCI UART peripheral driver for the KL25Z to support pinctrl, which should fix the CI failure. Can you retry CI on that board?

hakehuang commented:
> have updated the LPSCI UART peripheral driver for the KL25Z to support pinctrl, which should fix the CI failure. Can you retry CI on that board?

running.

danieldegrasse commented:
#42238 is merged, so this PR should be ready now.

gmarull commented on drivers/pwm/pwm_mcux_ftm.c:
```diff
@@ -441,6 +449,9 @@ static const struct pwm_driver_api mcux_ftm_driver_api = {
 
 #define TO_FTM_PRESCALE_DIVIDE(val) _DO_CONCAT(kFTM_Prescale_Divide_, val)
 
+#define PINCTRL_DEFINE(n) PINCTRL_DT_INST_DEFINE(n);
+#define PINCTRL_INIT(n) .pincfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),
```
why do we need this in this case? This driver seems to be using pinctrl unconditionally.

danieldegrasse replied:
We don't, I missed this. I've dropped the defined and used the macros directly, thanks for noticing.

gmarull commented on boards/arm/twr_ke18f/pinmux.c:
```diff
@@ -43,35 +44,121 @@ static int twr_ke18f_pinmux_init(const struct device *dev)
-	pinmux_pin_set(portd, 3, PORT_PCR_MUX(kPORT_MuxAlt3));
+	/* Configuation for SPI0 PCS2 KConfig */
+#if CONFIG_BOARD_TWR_KE18F_SPI_0_PCS2 && CONFIG_SPI
+	/* Struct is declared as extern here, LPSPI driver implements it */
+	extern struct pinctrl_dev_config
+		Z_PINCTRL_DEV_CONFIG_NAME(DT_NODELABEL(lpspi0));
```
please use `PINCTRL_DT_DEV_CONFIG_DECLARE`. See `samples/boards/nrf/dynamic_pinctrl` on how to use dynamic pinctrl.

danieldegrasse replied:
Fixed, thank you for the pointer

gmarull commented on boards/arm/twr_ke18f/pinmux.c:
```diff
@@ -43,35 +44,121 @@ static int twr_ke18f_pinmux_init(const struct device *dev)
+	CONFIG_BOARD_TWR_KE18F_SPI_1_PCS2 &&		\
+	CONFIG_SPI
+	/* Struct is declared as extern here, LPSPI driver implements it */
+	extern struct pinctrl_dev_config
+		Z_PINCTRL_DEV_CONFIG_NAME(DT_NODELABEL(lpspi1));
```
ditto

danieldegrasse replied:
fixed

gmarull commented on boards/arm/twr_ke18f/pinmux.c:
```diff
@@ -43,35 +44,121 @@ static int twr_ke18f_pinmux_init(const struct device *dev)
+	}
+#elif CONFIG_BOARD_TWR_KE18F_SPI_1_PCS0 && CONFIG_SPI
+	/* Struct is declared as extern here, LPSPI driver implements it */
+	extern struct pinctrl_dev_config
+		Z_PINCTRL_DEV_CONFIG_NAME(DT_NODELABEL(lpspi1));
```
ditto

danieldegrasse replied:
fixed

gmarull commented on boards/arm/twr_ke18f/pinmux.c:
```diff
@@ -43,35 +44,121 @@ static int twr_ke18f_pinmux_init(const struct device *dev)
+	}
+#elif CONFIG_BOARD_TWR_KE18F_SPI_1_PCS2 && CONFIG_SPI
+	/* Struct is declared as extern here, LPSPI driver implements it */
+	extern struct pinctrl_dev_config
+		Z_PINCTRL_DEV_CONFIG_NAME(DT_NODELABEL(lpspi1));
```
ditto

danieldegrasse replied:
fixed

gmarull commented on boards/arm/twr_ke18f/twr_ke18f.dts:
```diff
@@ -15,6 +15,17 @@
 	model = "NXP Kinetis KE18 MCU Tower System Module";
 	compatible = "nxp,mke18f16", "nxp,ke18f", "nxp,ke1xf";
 
+	/* Additional pinctrl states for flexio, as well as SPI PCS pins */
+	zephyr,user {
+		/* FlexIO clockout pin */
+		flexio_clockout_mux = <&flexio_clockout>;
+		/* LPSPI PCS muxes */
+		lpspi0_pcs2_mux = <&lpspi0_pcs2>;
+		lpspi1_pcs0_mux = <&lpspi1_pcs0>;
+		lpspi1_pcs2_mux = <&lpspi1_pcs2>;
+		lpspi1_pcs0_pcs2_mux = <&lpspi1_pcs0_pcs2>;
+	};
```
note: I'm not familiar with this board, but maybe you can provide overlays that select different choices instead of using dynamic pinctrl (or maybe board variants?). Dynamic pinctrl is used when the selection is not known beforehand, e.g. depends on input like a GPIO state.

danieldegrasse replied:
I think overlays would be the best option here. I know @henrikbrixandersen is relatively familiar with this board. @henrikbrixandersen, I'll update this PR to use overlays, and if that does not seem like an acceptable solution to you let me know and we can adjust.

danieldegrasse replied:
For the FlexIO clock pin, what would be the appropriate way to apply the pinmux? There is no upstream driver for NXP's FlexIO peripheral, but it does not fit the use case of dynamic pinctrl.

henrikbrixandersen replied:
Overlays sounds good, but please do not spend too much time on it. I can always fix it up later or use local overlays as needed. I doubt anybody else are using the board specific pinctrl Kconfig options.

danieldegrasse replied:
After playing with overlays for a bit, I think we might need to stick to dynamic pinctrl. The search directory list for devicetree overlays doesn't include the board directory, and even if it did the board `CMakeLists.txt` is parsed too late in the build process to add a devicetree overlay.

gmarull replied:
Btw, an example of a board that provides some extra overlays: https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/arm/particle_xenon/dts

danieldegrasse replied:
Thanks for the example, I've got overlays working by moving them to a `dts` folder in the board directory. The flexIO pin still is enabled using dynamic pinctrl. Generally speaking, is there a way to apply pins outside of a driver definition without using dynamic pin control?

gmarull replied:
You can do one trick to achieve that (haven't tried, but I think it should work):

1. Create a board level binding at `${BOARD}/dts/bindings/` (see e.g. `nrf9160dk_nrf52840`), if it's not board specific maybe just put it under `dts/bindings`.
2. Define your flexIO entry in DT like this:
```
flexio: flexio {
    compatible = "....";
    pinctrl-0 = <&flexio_default>; /* add pin config in the `-pinctrl.dtsi` file */
    pinctrl-names = "default";
};
```
3. In your board.c you can then do:
```c
#define FLEXIO DT_NODELABEL(flexio)
PINCTRL_DT_DEFINE(FLEXIO);
ret = pinctrl_apply_state(PINCTRL_DT_DEV_CONFIG_GET(FLEXIO), PINCTRL_STATE_DEFAULT);
...
```

gmarull reviewed (changes requested):
Overall LGTM, I have a few requests

danieldegrasse committed:
boards: mimxrt1060: add all pinctrl nodes for RT1060

all pinctrl nodes for RT1060 need to be defined, as some drivers used by
kinetis also are used on RT series boards.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
Just a note: I've added a few pinctrl definitions to the RT1060 in this PR, because some peripheral drivers are shared between RT and Kinetis SOCs. For these peripherals, the RT1060 needs to define pinctrl nodes, or the device will fail to initialize.

danieldegrasse committed:
boards: twr_ke18f: add flexio pinmux and binding

Add dts binding and pinctrl definition for flexio in twr_ke18f. This
allows ke18f pinmux code to apply pinctrl selections for flexio.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

gmarull reviewed (approved):
LGTM, great work
